flink-connectors/flink-connector-aws-base/src/test/java/org/apache/flink/connector/aws/testutils/AWSServicesTestUtils.java

```diff
@@ -97,77 +94,51 @@ public static Properties createConfig(String endpoint) {
         CreateRoleRequest request = CreateRoleRequest.builder().roleName(roleName).build();
 
-        CompletableFuture<CreateRoleResponse> responseFuture = iam.createRole(request);
-        responseFuture.get();
+        iam.createRole(request);
     }
 
-    public static List<S3Object> listBucketObjects(S3AsyncClient s3, String bucketName)
-            throws ExecutionException, InterruptedException {
-        ListObjectsRequest listObjects = ListObjectsRequest.builder().bucket(bucketName).build();
-        CompletableFuture<ListObjectsResponse> res = s3.listObjects(listObjects);
-        return res.get().contents();
+    public static List<S3Object> listBucketObjects(S3Client s3, String bucketName) {
+        ListObjectsV2Request listObjects =
+                ListObjectsV2Request.builder().bucket(bucketName).build();
+        ListObjectsV2Response res = s3.listObjectsV2(listObjects);
+        return res.contents();
     }
 
     public static <T> List<T> readObjectsFromS3Bucket(
-            S3AsyncClient s3AsyncClient,
+            S3Client s3Client,
             List<S3Object> objects,
             String bucketName,
             Function<ResponseBytes<GetObjectResponse>, T> deserializer) {
-        S3BucketReader bucketReader = new S3BucketReader(s3AsyncClient, bucketName);
-        return bucketReader.readObjects(objects, deserializer);
-    }
-
-    /** Helper class to read objects from S3. */
-    private static class S3BucketReader {
-        private final S3AsyncClient s3AsyncClient;
-        private final String bucketName;
-
-        public S3BucketReader(S3AsyncClient s3AsyncClient, String bucketName) {
-            this.s3AsyncClient = s3AsyncClient;
-            this.bucketName = bucketName;
-        }
-
-        public <T> List<T> readObjects(
-                List<S3Object> objectList,
-                Function<ResponseBytes<GetObjectResponse>, T> deserializer) {
-            return objectList.stream()
-                    .map(object -> readObjectWitKey(object.key(), deserializer))
-                    .collect(Collectors.toList());
-        }
-
-        public <T> T readObjectWitKey(
-                String key, Function<ResponseBytes<GetObjectResponse>, T> deserializer) {
-            GetObjectRequest getObjectRequest =
-                    GetObjectRequest.builder().bucket(bucketName).key(key).build();
-            return s3AsyncClient
-                    .getObject(getObjectRequest, AsyncResponseTransformer.toBytes())
-                    .thenApply(deserializer)
-                    .join();
+        List<T> results = new ArrayList<>();
+        for (S3Object object : objects) {
+            GetObjectRequest request =
+                    GetObjectRequest.builder().bucket(bucketName).key(object.key()).build();
+            ResponseBytes<GetObjectResponse> objectAsBytes = s3Client.getObjectAsBytes(request);
+            results.add(deserializer.apply(objectAsBytes));
         }
+        return results;
```

Comment:
I think the previous use of `stream()` was a bit cleaner.

We could do something like:
```
GetObjectRequest.Builder getObjectRequestBuilder = GetObjectRequest.builder().bucket(bucketName);
return objects.stream()
              .map(object -> getObjectRequestBuilder.key(object.key()).build())
              .map(s3Client::getObjectAsBytes)
              .map(deserializer::apply)
              .collect(Collectors.toList());
```

What do you think?

Reply:
Sounds like a great idea, will implement!